feat: implemented NIO scheduler and related tests

parvit opened the pull request:
/claim 9356

# Description
The following changes solve issue #9356 by implementing the NIO Scheduler algorithm found at  [Announcing Nio](https://nurmohammed840.github.io/posts/announcing-nio/).

Please review and give feedback.

### High-Level Overview

The scheduler operates on a fixed pool of worker threads and when a new task is created, the scheduler instance decides which worker thread is the best candidate to execute it. Its strategy is to send the task to the worker that currently has the fewest tasks in its queue.

However, to avoid the high cost and contention of checking every worker's queue length for every new task, NIO implements an optimization: it only checks a small, rotating subset of workers. This provides excellent load balancing with minimal overhead.

Once a task is assigned to a worker, it stays with that worker. If a task yields, it is placed in a local queue belonging to the same worker. This ensures that when the task is ready to run again, it's likely to be executed on the same CPU core, which is highly beneficial for cache performance.

## Pseudo-Code

This is a conceptual pseudocode analogue of the scheduler logic:

```plaintext
procedure Scheduler.schedule_new_task(task)
    /*
     * Select a small, rotating batch of workers to reduce
     * contention and search space.
     */
    batch = select_next_worker_batch()

    /*
     * Find the worker with the fewest queued tasks
     * within the selected batch.
     */
    least_loaded_worker = find_least_loaded_in_batch(batch)

    /*
     * Dispatch the new task to the chosen worker's main queue.
     * This is the "cold path" for a task.
     */
    least_loaded_worker.main_queue.push(task)

procedure Worker.execution_loop()
    loop
        /*
         * HOT PATH: Prioritize the local queue containing tasks
         * that have yielded on this worker before. This maximizes
         * CPU cache reuse.
         */
        task = local_defer_queue.pop_front()

        if task is None then
            /*
             * COLD PATH: If the hot path is clear, check for new
             * tasks from the scheduler.
             */
            task = main_queue.try_receive()

        if task is None then
            /*
             * If both queues are empty, block until the scheduler
             * sends a new task.
             */
            task = main_queue.blocking_receive()

        if task is SHUTDOWN then
            break

        has_yielded = task.process()

        if has_yielded then
            /*
             * If the task is not complete, return it to the local
             * queue to keep it on the hot path.
             */
            local_defer_queue.push_back(task)
        else
            /*
             * The task is complete and can be dropped.
             */
            drop(task)
```


# Implemented classes

* **NIOExecutor** : The `Executor` class that implements the NIO algorithm.

* **NIOClock** : A ZIO `Clock` implementation that delegates all time-based operations like `sleep` directly to the `NIOExecutor`'s internal timer for integrated scheduling.

# Available configurations

The behavior of the NIOExecutor can be configured by three parameters:

- **nThreads**  : the number of threads to allocate to the scheduler's pool, defaults at the number of processors
- **batchSize** : the size of the batch of workers to sample to check the current load, defaults to 4
- **trace**     : activates the debug trace logging of the scheduler, defaults to false


# Performance

To evaluate the performances of the new algorithm, benchmarks were implemented based on the ones already present for ZScheduler, to allow comparing the effectiveness of the new algorithm across the same types of workload.

The tests were executed on a github actions standard `ubuntu-latest` instance so that the experiments are replicable, attached to this PR find the log file and workflow used for the test (it is not part of the PR as it was purpose created and not as a generic benchmark workflow).


### Benchmark Data

Following the relevant data 

**NIOExecutorBenchmarks**
```
NIOExecutorBenchmarks.nioExecutorChainedFork         thrpt   45  4468.817 ± 14.008  ops/s
NIOExecutorBenchmarks.nioExecutorForkMany            thrpt   45   416.154 ±  4.132  ops/s
NIOExecutorBenchmarks.nioExecutorPingPong            thrpt   45  1173.240 ± 14.806  ops/s
NIOExecutorBenchmarks.nioExecutorYieldMany           thrpt   45    69.935 ±  0.444  ops/s
NIOExecutorBenchmarks.zioFixedThreadPoolChainedFork  thrpt   45  2399.341 ± 21.308  ops/s
NIOExecutorBenchmarks.zioFixedThreadPoolForkMany     thrpt   45   273.567 ±  8.079  ops/s
NIOExecutorBenchmarks.zioFixedThreadPoolPingPong     thrpt   45   785.472 ±  7.855  ops/s
NIOExecutorBenchmarks.zioFixedThreadPoolYieldMany    thrpt   45    25.160 ±  0.393  ops/s
NIOExecutorBenchmarks.zioSchedulerChainedFork        thrpt   45  2144.533 ± 35.478  ops/s
NIOExecutorBenchmarks.zioSchedulerForkMany           thrpt   45   303.918 ±  2.486  ops/s
NIOExecutorBenchmarks.zioSchedulerPingPong           thrpt   45  1762.219 ± 12.856  ops/s
NIOExecutorBenchmarks.zioSchedulerYieldMany          thrpt   45   123.128 ±  4.117  ops/s
```

**ZSchedulerBenchmarks**
```
ZSchedulerBenchmarks.catsRuntimeChainedFork         thrpt   45  3073.415 ± 836.250  ops/s
ZSchedulerBenchmarks.catsRuntimeForkMany            thrpt   45   494.755 ±   2.423  ops/s
ZSchedulerBenchmarks.catsRuntimePingPong            thrpt   45   984.364 ±   5.746  ops/s
ZSchedulerBenchmarks.catsRuntimeYieldMany           thrpt   45   170.957 ±   2.317  ops/s
ZSchedulerBenchmarks.zioFixedThreadPoolChainedFork  thrpt   45  2289.520 ±  37.753  ops/s
ZSchedulerBenchmarks.zioFixedThreadPoolForkMany     thrpt   45   273.829 ±   2.704  ops/s
ZSchedulerBenchmarks.zioFixedThreadPoolPingPong     thrpt   45   808.391 ±   3.676  ops/s
ZSchedulerBenchmarks.zioFixedThreadPoolYieldMany    thrpt   45    24.290 ±   0.211  ops/s
ZSchedulerBenchmarks.zioSchedulerChainedFork        thrpt   45  2257.668 ±  22.076  ops/s
ZSchedulerBenchmarks.zioSchedulerForkMany           thrpt   45   306.461 ±   3.363  ops/s
ZSchedulerBenchmarks.zioSchedulerPingPong           thrpt   45  1743.525 ±  13.358  ops/s
ZSchedulerBenchmarks.zioSchedulerYieldMany          thrpt   45   122.391 ±   8.624  ops/s
```

**CustomExecutorBenchmark**
```
CustomExecutorBenchmark.customExecutor                20  thrpt   10  3221.490 ± 34.020  ops/s
CustomExecutorBenchmark.customExecutorInitBoth        20  thrpt   10  3185.611 ± 25.067  ops/s
CustomExecutorBenchmark.defaultExecutor               20  thrpt   10  3905.783 ± 24.350  ops/s
CustomExecutorBenchmark.defaultExecutorInitBoth       20  thrpt   10  3266.086 ± 64.701  ops/s
CustomExecutorBenchmark.nioExecutor                   20  thrpt   10  2984.689 ± 34.566  ops/s
CustomExecutorBenchmark.nioExecutorInitBoth           20  thrpt   10  3072.759 ± 20.743  ops/s
```

Based on the attached results, the new scheduler demonstrates strong performance in various scenarios, particularly those involving rapid task creation and chaining, but shows some weaknesses in others, especially under heavy CPU contention.

parvit committed:
feat: implemented NIO scheduler and related tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zio/zio?pullRequest=10220) <br/>All committers have signed the CLA.

parvit commented:
[benchmark-log.zip](https://github.com/user-attachments/files/23382716/benchmark-log.zip)
[benchmarks.yml](https://github.com/user-attachments/files/23382717/benchmarks.yml)

parvit committed:
applied formatting rules and fixed warning

parvit committed:
applied formatting rules

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

kyri-petrou commented:
@parvit I will review the PR once CI is passing :)

parvit commented:
@kyri-petrou the CI is failing for timeout of 1 hour... could you shed some light on this problem? if it is related to my change i'll be happy to correct it in any way i can.

kyri-petrou commented:
@parvit AFAICT, the problem is that one of the specs that you added is modifying the executor, and for some reason that executor is being used in other suites as well.

The concerning part is that because the NIOExecutor is being used in other specs, it's causing deadlocks which means that the executor implementation is either not thread-safe or that it has some bug in it.

To approve this PR, I would like to see all the tests running with the NIOExecutor and CI passing. To do this we'd need to make the executor configurable via either envvars or a property, and add a new CI job that runs the tests using the NIOExecutor.

Before you go into a rabbithole of trying to set up CI and stuff, please make sure that everything is passing locally using the NIOExecutor in all tests

parvit committed:
fix hanging tests issue

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

parvit commented:
@kyri-petrou Hi, now the CI is passing, please review.

The issue was relative to using `Runtime.setExecutor` inside the tests which changed the executor for all tests and lead to the hanging issue.

Now the `NIOExecutorSpec` tests are launched using `ZIO.serviceWithZIO` which overrides the default executor with NIOExecutor only for the lifetime of the test.

kyri-petrou commented:
@parvit I think the fix that you did is only masking the issue. We need to ensure that all of ZIO's tests pass when using the `NIOExecutor`.

parvit commented:
@kyri-petrou I see. I'll investigate how to implement that and let you know.

parvit committed:
configurable test executor

parvit committed:
Merge branch 'dev-nio-executor' of https://github.com/parvit/zio into dev-nio-executor

parvit committed:
fixed NIOExecutor non-locking behavior

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

parvit commented:
@kyri-petrou i've added the support for specifying on commandline the parameter "zio.test.executor" which can have the values: "nio", "zscheduler" and "default".

It is able to override the default executor used for all tests like you requested.

I've also updated the CI.yml to execute the tests with the two parameters so that it can be checked.

parvit committed:
applied formatting

parvit committed:
implemented override for default executor in runtime

parvit committed:
applied formatting, added runtime test files for js

parvit committed:
moved config test files to support all configurations

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

parvit commented:
@kyri-petrou With the last pull from the series branch now the entire CI process passes with both executors, i'll wait for your review now.

parvit committed:
Merge branch 'series/2.x' into dev-nio-executor

parvit commented:
@hearnadam why have you closed this? it is waiting for review from @kyri-petrou. 
I've added everything that was requested so i don't see the motivation of the sudden closure.

hearnadam commented:
@parvit I closed the PR as it's unclear to me that you have addressed Kyri's comments. Additionally, aspects of the code don't seem to be particularly well suited to ZIO, but instead copy/pasted or AI generated. For example: `TraceLogger` using a non-volatile.

parvit commented:
@hearnadam Yes i've had help from AI as i do not have a lot of experience with Scala syntax and ZIO but that does not mean i did not address the issues? i've added the tests execution on the CI for all the executors so that it shows that there are no regressions.

If you have additional concerns we can address them but closing it all i think its not the right approach.
